### PR TITLE
Add `Room.getLastLiveEvent` and `Room.getLastThread`

### PR DESCRIPTION
**Pull request by weeman1337**
Part of https://github.com/vector-im/element-web/issues/25207

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

Notes: Added `Room.getLastLiveEvent` and `Room.getLastThread`. Deprecated `Room.lastThread` in favour of `Room.getLastThread`.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Added `Room.getLastLiveEvent` and `Room.getLastThread`. Deprecated `Room.lastThread` in favour of `Room.getLastThread`. ([\#3321](https://github.com/matrix-org/matrix-js-sdk/pull/3321)).

### CODE REVIEW

File: src/models/room.ts [hunk 1]
```diff
@@ -396,7 +396,6 @@ export class Room extends ReadReceipt<RoomEmittedEvents, RoomEventHandlerMap> {
      * This is not a comprehensive list of the threads that exist in this room
      */
     private threads = new Map<string, Thread>();
-    public lastThread?: Thread;
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply src/models/room.ts:399]: This is a breaking change and will require yet another major version bump.

[the remark this answers is not in the document]
Comment (andybalaam) [reply src/models/room.ts:399]: Which is absolutely worth it if it helps fix the stuck unreads problem. We should batch together breaking changes if we can.

[the remark this answers is not in the document]
Comment (t3chguy) [reply src/models/room.ts:399]: IMO it should be kept & deprecated even if it isn't used by Element

[the remark this answers is not in the document]
Comment (weeman1337) [reply src/models/room.ts:399]: We can keep `lastThread` with a deprecation note (\+ a comment mentioning that it does not contain a reliable value). Then we can remove it together with other breaking changes soon.

[the remark this answers is not in the document]
Comment (t3chguy) [reply src/models/room.ts:399]: Perfect

[the remark this answers is not in the document]
Comment (andybalaam) [reply src/models/room.ts:399]: Sounds good.

File: src/models/room.ts [hunk 2]
```diff
@@ -785,6 +784,32 @@ export class Room extends ReadReceipt<RoomEmittedEvents, RoomEventHandlerMap> {
...
+     */
+    public getLastLiveEvent(): MatrixEvent | undefined {
+        const roomEvents = this.getLiveTimeline().getEvents();
+        const lastRoomEvent = roomEvents[roomEvents.length - 1] as MatrixEvent | undefined;
+
+        return this.getThreads().reduce<MatrixEvent | undefined>(
+            (lastEventSoFar: MatrixEvent | undefined, thread: Thread) => {
+                const lastThreadEvent = thread.events[thread.events.length - 1];
+
+                if (!lastEventSoFar || (lastThreadEvent?.getTs() ?? 0) > (lastEventSoFar?.getTs() ?? 0)) {
```
Comment (andybalaam) [inline src/models/room.ts:802]: Please can you add a comment here and everywhere else you know of that uses `ts` for event ordering, referring to https://github.com/matrix-org/matrix-js-sdk/issues/3325 ?

File: src/models/room.ts [hunk 3]
```diff
@@ -785,6 +789,45 @@ export class Room extends ReadReceipt<RoomEmittedEvents, RoomEventHandlerMap> {
...
+
+        const lastThreadEvent = lastThread.events[lastThread.events.length - 1];
+
+        return (lastRoomEvent?.getTs() ?? 0) > (lastThreadEvent.getTs() ?? 0) ? lastRoomEvent : lastThreadEvent;
+    }
+
+    /**
+     * @returns the thread with the most recent event in its live time line. undefined if there is no thread.
+     */
+    public getLastThread(): Thread | undefined {
```
Comment (t3chguy) [inline src/models/room.ts:814]: Could this not just become `get lastThread` and thus not need a deprecation or any downstream changes?

Comment (t3chguy) [reply src/models/room.ts:814]: Oh I guess the existing one is externally writable so no

File: src/models/room.ts (hunk 2 above)
Comment (weeman1337) [reply src/models/room.ts]: Done

**Comment by weeman1337** [comment]
> SonarCloud Code Analysis Failing after 1m — Quality Gate failed 

image

→ Ignore SonarCloud in this case?

**Review by t3chguy — APPROVED** [review]
:shipit:

**Pull request merged**